testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java

```diff
@@ -0,0 +1,36 @@
+package edu.illinois.cs.testrunner.execution;
+
+import edu.illinois.cs.testrunner.data.results.TestRunResult;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.junit.platform.engine.discovery.DiscoverySelectors;
+import org.junit.platform.engine.discovery.MethodSelector;
+import org.junit.platform.launcher.Launcher;
+import org.junit.platform.launcher.LauncherDiscoveryRequest;
+import org.junit.platform.launcher.core.LauncherDiscoveryRequestBuilder;
+import org.junit.platform.launcher.core.LauncherFactory;
+
+// Use JUnit 5 API to run tests, including JUnit 4 and 5 tests.
+public class JUnit5TestExecutor {
+
+    public static TestRunResult runTestsSeparately(
```

Comment:
We discussed an improvement that seems to be missing from this PR (but can be added later): when running tests separately, consecutive tests from the same test class would still be run together (ClassA.t1, ClassA.t2, ClassB.t3, ClassA.t4 would only invoke the launcher 3 times and not 4 times) 

Where are these potential improvements being tracked at the moment?

Reply:
Yea, I will add a TODO here.

Reply:
Why is the method called runTestsSeparately while in JUnitTestExecutor the corresponding method seems to be runOrder? Is there something special about JUnit 5 such that it needs to be called runTestsSeparately?

Reply:
There was a long conversation about this. In short, I haven't found a way to execute tests with full control of the order, if we don't run "separately". This is a workaround so that we have full control of the execution order. Refer [here](https://github.com/luoos/junit5-testrunner-demo/blob/1d0e4388b1193902eac4b227c37c12fc0e31eae9/src/main/java/com/luojl/demo/TestRunner.java#L17) to some details.

Reply:
IIRC the bigger problem we have wouldn't exactly be how these JUnit 5 tests are being run (i.e., everytime a test is run, it's @BeforeClass and @AfterClass would run), but that this behavior would be inconsistent with what we are doing for JUnit 4 tests (i.e., a test class's @BeforeClass is only run once, the very first time a test method from the class is run even if tests from other classes interleave).

As I don't see strong reasons to choose one behavior over the other, we should just look to include making the way JUnit 4 and JUnit 5 treat @BeforeClass/@AfterClass consistent in our list of TODOs/Caveats.

Reply:
Done. Add some annotation here.

Reply:
I don't understand what you want a reader of the TODO to do. The problem as I said is mainly that the behavior for JUnit 4 and JUnit 5 is different in testrunner now.

The current way that JUnit 5 tests are run doesn't seem problematic to me. If it is hard (or impossible) to replicate JUnit 4's behavior in JUnit 5, shouldn't this TODO just tell the reader to implement JUnit 5's behavior for JUnit 4?

The text you have now largely explains why it is difficult to implement JUnit 4's behavior in JUnit 5, which is fine to keep. However, I don't know why "optimal solution" is to replicate JUnit 4's behavior in JUnit 5. That to me, seems to be a solution, but as you pointed out is hard or impossible of a solution anyway.

Reply:
I also don't see any mentioning of the TODO to add the feature of "when running tests separately, consecutive tests from the same test class would still be run together and ``@BeforeClass` and `@AfterClass` would only be run once for them".

Overall, I see two TODOs when we merge your current changes:
(1) the feature of "when running tests separately, consecutive tests from the same test class would still be run together and ``@BeforeClass` and `@AfterClass` would only be run once for them"
(2) make JUnit 4 and JUnit 5's behavior consistent in how it treats `@BeforeClass` and `@AfterClass` (you can then clarify for this TODO how one of the solutions (making JUnit 5 be like JUnit 4) is hard or impossible

Can you please clarify the two TODOs here?

Reply:
Let's step back a little bit, and think about a more fundamental question. The reason for the "separate" method is to get **full** control of the test order. But do we really need that?

The JUnit 5 API already provides a simple API to get the **almost-full** control of the test order. There are just two exceptions: 1) interleaved class 2) Order annotation.

However, the above two situations will never happen in the real case.

1) **Interleaved class**: When a user runs a test suite, it's impossible to run some tests in ClassA, then some in ClassB, then again some in ClassA. Even if iDFlakies can run tests in a class-interleaved way, the "flaky" tests found in this way are just a false alert.
2) **Order annotation**: Users already specify the order for their tests. If we force to change the order and find some "flaky" tests, that would be a false alert as well.

All in all, my point is that we don't need the "separate" method to get the **full** control. Using the simple JUnit 5 API (like [this](https://github.com/luoos/junit5-testrunner-demo/blob/1d0e4388b1193902eac4b227c37c12fc0e31eae9/src/main/java/com/luojl/demo/TestRunner.java#L15)) to get the "**almost-full**" control is good enough. In this way, there is no performance overhead and @BeforeClass-things will behave like JUnit 4.

Based on the above behavior, we need to modify the shuffling logic: we need to make sure 1) there is no class-interleaved case, and 2) don't shuffle tests with Order annotation.

@winglam what do you think?

Reply:
Are you suggesting that instead of merging `runTestsSeparately` you would rather just merge an implementation that grants you almost-full control of the test order? If so, why not strive to have both at some point and you can just merge `runTestsSeparately` now? (IIRC you already had both implementations ready the last time we met and Darko's suggestion was just to add a configuration option to choose between the two.)

Your concern with `runTestsSeparately` is largely centered around the fact that it creates "false alerts". Indeed, if one is using `runTestsSeparately` to detect OD tests, then the detected tests may be construed as false positives to the developers. However, imagine iFixFlakies trying to find a cleaner. In such a case, it doesn't matter that the cleaner requires the interleaving of two classes or violates the order annotation.

As I have been saying, there isn't a strong reason that one implementation is strictly better than another as each implementation will benefit one use case or another. The most problematic part of your changes here is simply that the current implementation of testrunner will handle JUnit 4 and JUnit 5 tests differently (i.e., JUnit 4 runs `@BeforeClass` once for every test class no matter the order, while JUnit5 may run `@BeforeClass` many times). For now, it suffices to document this problem as a TODO so that when you or others have more time, it can be addressed then.

Reply:
Sounds reasonable with the point related to iFixFlakies. OK, let's try to merge this "separate" one first. I will modify the TODO to highlight the different behavior between J4 and J5.

Reply:
Updated the annotation. Please resolve this conversation if it looks good to you.

Reply:
Thanks for adding the detailed information for the TODO related to JUnit 4 and JUnit 5 differences.

Can you please add a TODO for the following with the relevant details as well?
> the feature of "when running tests separately, consecutive tests from the same test class would still be run together and `@BeforeClass` and `@AfterClass` would only be run once for them"

Reply:
Updated. But I think the effect of that improvement would be limited, especially when there are thousands of tests. In that case, the possibility to find consecutive and same-class tests is pretty low.